PyComPWA.cpp

```diff
@@ -173,7 +173,48 @@ PYBIND11_MODULE(ui, m) {
       .def(py::init<std::vector<ComPWA::pid>, std::vector<ComPWA::Event>>(),
            py::arg("pids"), py::arg("events"))
       .def_readonly("pids", &ComPWA::EventCollection::Pids)
-      .def_readonly("events", &ComPWA::EventCollection::Events);
+      .def_readonly("events", &ComPWA::EventCollection::Events)
+      .def("to_table",
+           [](const ComPWA::EventCollection &self) {
+             if (!self.checkPidMatchesEvents())
+               throw ComPWA::CorruptFile(
+                   "Number of PIDs in EventCollection does not match the "
+                   "number of four momenta in each of its events");
+             std::vector<std::vector<double>> table;
```

Comment:
I would define it as a `unordered_map<string, vector<double>>`. which has an easier memory memory layout and can be converted to a pandas dataframe easily as well

Reply:
I gave it a try just now, but an `unordered_map` (bound to a `dict` with `str` as keys in python) won't allow to create a `DataFrame` with multicolumns. For that, the `data` would have to be a `dict` with 2-tuples of strings as keys, like this:

```python
import pandas as pd

dummy = np.zeros((5,))
input_data = {
    ('gamma', 'p_x'): dummy,
    ('gamma', 'p_y'): dummy,
    ('gamma', 'p_z'): dummy,
    ('gamma', 'E'): dummy,
    ('pi0', 'p_x'): dummy,
    ('pi0', 'p_y'): dummy,
    ('pi0', 'p_z'): dummy,
    ('pi0', 'E'): dummy,
}
pd.DataFrame(input_data)
```

So I like the idea of an `unordered_map`, remember now that the only way to create such a `DataFrame` structure is to create a [`MultiIndex` column](https://pandas.pydata.org/pandas-docs/stable/user_guide/advanced.html) and feed it a 'block' of data. Hence the vector of a vector.

Reply:
Hmm ok. Then `unordered_map<tuple<string, string>, vector<double>>` would be the way to go. Unfortunately you would have to provide a custom hashing function for the `tuple`/`pair` of strings... (you could reuse some code in SubSystem.hpp, but it seems overkill for this).
You could use a `map<pair<string, string>, vector<double>>`, though. This should work.

Reply:
Tried with the following just now:

```cpp
.def("to_table",
      [](const ComPWA::EventCollection &self) {
        if (!self.checkPidMatchesEvents())
          throw ComPWA::CorruptFile(
              "Number of PIDs in EventCollection does not match the "
              "number of four momenta in each of its events");
        std::map<std::pair<int, std::string>, std::vector<double>> Table;
        for (size_t i = 0; i < self.Pids.size(); ++i) {
          auto Pid = self.Pids[i];
          auto KeyPx = std::make_pair(Pid, "p_x");
          auto KeyPy = std::make_pair(Pid, "p_y");
          auto KeyPz = std::make_pair(Pid, "p_z");
          auto KeyE = std::make_pair(Pid, "E");
          for (const auto &Event : self.Events) {
            Table[KeyPx].push_back(Event.FourMomenta[i].px());
            Table[KeyPy].push_back(Event.FourMomenta[i].py());
            Table[KeyPz].push_back(Event.FourMomenta[i].pz());
            Table[KeyE].push_back(Event.FourMomenta[i].e());
          }
        }
        return Table;
      })
```

But there is a big problem: the PIDs in the final states are not always be unique, so this map won't result in the right form of table. This is why the `data.naming` module was necessary.

Reply:
But can't you integrate the pid unique feature here?
To me there are these use cases:
1. convert an `EventCollection` to a `DataFrame`. 
Here I don't understand why parts of the work is done in the `to_table` and the other in a `convert` function. Can't you just return either a correct `map<>` or do everything in the `convert`. (I can't judge atm which one is the better way)
2. convert a `DataSet` to a `DataFrame`.
Conversion is trivial, but some argument applies as above. Here the hardest part is changing the unique ids to a particle name.

Or am I missing something?

Reply:
May be worth a try, but that would mean writing something like `naming.make_values_unique` in C++ (was already a pain in Python). That one appends counter indices to duplicate values in a list (in this case, the keys of the map). I didn't want to go into that within that Python binding definition, but can try if you think it is worth pursuing.

Reply:
Hmm, I have to think about it for a bit. First we need to decide if it should be done in python or c++. In the current implementation things are mixed together.

Reply:
As the C++/Python binding in the current form will probably be abandoned, I think it's better not to get too lost in this and keep the renaming procedure on the Python side. The mixing isn't clean, but this is also an artifact of the current setup.

Reply:
Ok, I think it would make more sense to do it on the python side. So my idea would be to remove `to_table`. And instead move this code to the `convert` function on the python side. It might be that this conversion on the python side is too slow though. Then we can just leave things as they are. What do you think? I also don't want you to overengineer this things since it will be archived later on.
To me this again suggest that the Event structure is bad design and is not needed. Instead the `EventCollection` should be stored in this table format from the beginning. This would be the actual solution to all of the problems connected to this. Since this also would affect the generator interface, that an important consideration for the tensorwaves interface. ComPWA/tensorwaves#18

Reply:
All right